### ISSUE TITLE
recover from failed decode

### ISSUE THREAD

**Issue by nickdesaulniers** [google/shaka-player#215]
While live streaming, gpac's DashCast sometimes produces an invalid segment.  I'll file an issue with them, but it would be useful if the client could recover from a segment that failed to decode.

<img width="1440" alt="screen shot 2015-10-23 at 10 52 02 am 1" src="https://cloud.githubusercontent.com/assets/1823839/10700665/287d6278-7974-11e5-8e3e-9547457a1dab.png">

**Comment by joeyparrish** [comment]
How would the client do such a thing?  If the video element throws a decode error and the MSE implementation invalidates all the source buffers, it would seem there's not much a JS app can do.

**Comment by nickdesaulniers** [comment]
It seems poor that an MSE implementation would invalidate everything because of one source buffer having a decode error.  So I guess we should see if the decode error is thrown by append `SourceBuffer.prototype.appendBuffer` and if that can be caught/recovered from.

**Comment by nickdesaulniers** [comment]
As in, it seems like the logic "this one source buffer is bad, therefore, let's invalidate all previous valid buffers" seems incorrect.

**Comment by joeyparrish** [comment]
Is this something Firefox's MSE implementation can recover from?

**Comment by ddorwin** [comment]
[`appendBuffer()`](http://w3c.github.io/media-source/#widl-SourceBuffer-appendBuffer-void-ArrayBufferView-data) doesn't throw any errors or do any decoding. Decode errors occur in the media element when media data from the appended buffer is used.

Adding @wolenetz for any other comments.

**Comment by wolenetz** [comment]
@ddorwin, correct: the error state of HTMLMediaElement is extended by MSE to include similar sources of errors (media parsing and decoding) that would occur in a regular non-MSE src=URL playback.

Specifically, in MSE's end of stream algorithm, as run by the append error algorithm, which is run when the asynch processing of appended media encounters decode error:

"If the HTMLMediaElement.readyState attribute equals HAVE_NOTHING
* Run the "If the media data can be fetched but is found by inspection to be in an unsupported format, or can otherwise not be rendered at all" steps of the resource fetch algorithm.

If the HTMLMediaElement.readyState attribute is greater than HAVE_NOTHING
* Run the media data is corrupted steps of the resource fetch algorithm."

These trigger fatal decode error steps in the HTMLMediaElement. When similar happens in a src= playback, would further fetching, seeking, etc be expected to succeed without restarting the load algorithm?

**Comment by joeyparrish** [comment]
@wolenetz Thanks, makes sense.  This matches with my experience in testing and debugging, as well.

@nickdesaulniers So I'm not seeing any reasonable way for a player to recover from a fatal error in the media.  Besides, if the content is fundamentally broken, shouldn't that be a bug on the software creating the content?  Have you reported the bug to gpac?

**Comment by nickdesaulniers** [comment]
>  Besides, if the content is fundamentally broken, shouldn't that be a bug on the software creating the content?

Absolutely.  Could it ever be the case that only part of the file was transferred (maybe chunked encoding) that prevented proper decoding? 

> Have you reported the bug to gpac?

https://github.com/gpac/gpac/issues/297
It might be useful to log which URL/segment triggered the invalid state, but it sounds like the HTMLMediaElement does not provide this info.

**Comment by nickdesaulniers** [comment]
@cpearce, does this sound correct, that use case can't be supported?

**Comment by nickdesaulniers** [comment]
or maybe @jyavendard can add some thoughts?

**Comment by joeyparrish** [comment]
If only part of the file was transferred, the request failed and we don't append.

**Comment by wolenetz** [comment]
If you try in Chrome and obtain a failure repro, take a look at chrome://media-internals. Especially in more recent builds of chrome (beta or dev branches), there could be some informative logs available for the failed playback.

**Comment by ddorwin** [comment]
>If only part of the file was transferred, the request failed and we don't append.

Is this any more interesting with `appendStream()`? I assume that is more like the `.src=` case as far as the media element is concerned.

**Comment by wolenetz** [comment]
appendStream() has essentially invokes the segment parser loop in a loop (like looping appendBuffer() with minor changes to make sure updatestart/end happen at most once per appendStream()). As such, the spec language requires appendStream's loop to abort if the segment parser loop aborted (which happens just after it runs the append error algorithm, just like in appendBuffer()). Outside of the built-in loop and using ReadableStream instead of an ArrayBuffer[View], appendStream() differs from appendBuffer() in that the former can run the coded frame eviction algorithm more than once (and that any such runs within the asynch loop portion do not result in QuotaExceededErr, rather append error algorithm without decode error, if not enough frames could be evicted).

In short, if GC within appendStream()'s asynch loop can't evict enough to make room, the appendStream() is aborted, but the HTMLMediaElement's fetching/load algorithms aren't aborted. But if decode err indeed occurred within appendStream()'s asynch loop, the result should be very similar to appendBuffer() of same media hitting decode error.

**Comment by joeyparrish** [comment]
Let's not get off-topic.  To wrap up on side-tracks:

1) Shaka doesn't use `appendStream()` and has no plans to do so in the forseeable future.
2) If the content is bad (which I think we agree on), follow up with gpac on the issue you filed there.
3) If you take issue with how Chrome handles content it can't decode, please file at crbug.com.

And to close the main topic:

At the moment, it appears that content that can't be decoded is a fatal error from the browser's point of view, so Shaka Player can't reasonably do anything about it.  In any case, we do not, as a matter of policy, go out of our way to "fix" bad content in the first place.

**Comment by wolenetz** [comment]
Sounds good to me. @ddorwin, regarding appendStream() side-track: abort of appendStream() without decode error can additionally be reached if the stream gets into a bad state:

If stream.state equals "errored", then run the append error algorithm with the decode error parameter set to false and abort this algorithm.
If stream.state does not equal "readable", then run the append error algorithm with the decode error parameter set to false and abort this algorithm.
